Don't render deleted messages

andrii-i opened the pull request:
### Problem

Currently when message is deleted it is still rendered in UI as `(message deleted)`. It seems to not bring any useful information and increase visual clutter.

### Proposed solution 

Do not render deleted messages. 

#### Before this PR:
![Screenshot 2025-06-18 at 5 21 05 PM](https://github.com/user-attachments/assets/27a1e949-9a67-4ef8-921e-4e5c0700d216)
![Screenshot 2025-06-18 at 5 21 18 PM](https://github.com/user-attachments/assets/220ea62f-d9ea-4d50-ac97-028e6dea9c20)
![Screenshot 2025-06-18 at 5 22 10 PM](https://github.com/user-attachments/assets/d9ea77da-ea09-4f4f-9e5d-746a60574275)

#### After this PR:
![Screenshot 2025-06-18 at 5 24 29 PM](https://github.com/user-attachments/assets/b5764883-49bc-43c6-a666-c716785fc734)
![Screenshot 2025-06-18 at 5 24 33 PM](https://github.com/user-attachments/assets/efb3d5f9-258b-4a94-b778-28804d3aab3b)
![Screenshot 2025-06-18 at 5 25 03 PM](https://github.com/user-attachments/assets/254fc0bb-9a17-4b96-b841-aa0a928278a7)

andrii-i committed:
Don't render deleted messages, handle null render promises

andrii-i committed:
update delete message test

ellisonbg commented:
Nice! Could you create another screenshot to show the view before the messages are deleted?

andrii-i commented:
@ellisonbg it was already present in "Problem" section unless I misunderstand. I moved it to the "Proposed solution" section and added "Before this PR:" just before the screenshot for more visibility.

ellisonbg commented:
Clarification: I was hoping you could show the messages before the user hit the delete button to delete the messages in the first place. Looking for before and after the user hits delete, rather than only before and after the PR.

andrii-i commented:
@ellisonbg thank you for the clarification, updated screenshots based on it. Please let me know if it looks good now.

brichet commented:
Thanks @andrii-i for working on this.

I think this is a matter of opinion, and agree that in the context of a conversation with an AI it doesn't make sense to keep this track.
In a regular use of the chat, I always find it frustrating when I saw an information and cannot retrieve any track of it in the history (like in Slack I think). In most of the others chat that I know, the *message deleted* is kept (gitter, zulip, whatsapp...).

Since this PR does actually remove the message from the shared document but only avoid displaying it, I would add a user setting for it.

andrii-i commented on packages/jupyter-chat/src/components/messages/messages.tsx:
```diff
@@ -115,7 +115,12 @@ export function ChatMessages(props: BaseMessageProps): JSX.Element {
       if (!allRendered) {
-        Promise.all(renderedPromise.current.map(p => p.promise)).then(() => {
+        const activePromises = renderedPromise.current
+          // Filter out nulls signigying deleted messages
```
yes, thank you

andrii-i committed:
fix typo in a comment

andrii-i commented:
Thank you for looking into this @brichet. I agree that this is a matter of opinion, that both approaches have tradeoffs, and indeed believe that approach presented in this PR makes more sense. 

> In a regular use of the chat, I always find it frustrating when I saw an information and cannot retrieve any track of it in the history (like in Slack I think). In most of the others chat that I know, the message deleted is kept (gitter, zulip, whatsapp...).
>
> Since this PR doesn't actually remove the message from the shared document (only avoid displaying it), I would add a user setting for it.

Would you be open to merging this PR and then following up with another PR later on the settings? Another way to deal with being able to retrieve messages could be be wiring up undo/redo so users can undo deleted messages to recover them.

brichet commented:
@andrii-i what I meant was not being able to retrieve the message (currently the content is deleted anyway and I think it's fine), but rather being able to know that message has been deleted. The current behavior looks fine for me.
The settings I mentioned could be only a boolean, whether to display or not the deleted messages, to let users choose the behavior.

andrii-i commented:
@brichet makes sense, I've updated #237 description to track the need of adding rendering of deleted messages as a boolean option. Are we good to merge this PR and follow up with another PR on the settings?

brichet commented:
> @brichet makes sense, I've updated #237 description to track the need of adding rendering of deleted messages as a boolean setting. Are we good to merge this PR and follow up with another PR on the setting?

Why not adding this settings in this PR ? It doesn't seem to fix anything broken, it's only a matter of opinion...

ellisonbg commented:
I would like to think a bit more about this and do some reading in the UX/HCI literature. Looks like there are some subtle UX questions here. The existing research tends to focus mostly on human-human chat, and there are definitely new dimensions when you add AI. I have definitely run into situations where having the "This message was deleted" was really painful, but there are other situations where it may make more sense. It isn't obvious that the two modes are a per-person preference, but may be more contextual. Some other ideas to explore:

* We could handle it differently if the deleted message was an AI message versus your own human one.
* We could include information about who did the deleting "Brian deleted this message".
* We could show the "This message was deleted" for a certain amount of time and then have it go away for good.
* We could show the "This message was deleted" but have an additional interaction for dismissing that for good.
* We could collapse multiple consecutive "This message was deleted" into a single "These messages were deleted" to take up less space (@brichet as an initial start, are you open to this simplification?)

Cheers

brichet commented:
> * We could collapse multiple consecutive "This message was deleted" into a single "These messages were deleted" to take up less space

:+1: IMO it's a better user experience than the current one when consecutive messages are deleted.

> * We could handle it differently if the deleted message was an AI message versus your own human one.

I don't have a strong opinion on it, and maybe we could remove them from the document, if we think it doesn't bring any information.
